File: Difficulty: Medium/Job Sequencing Problem/job-sequencing-problem.cpp
#define pi pair<int,int>


class Solution {
  public:
    vector<int> jobSequencing(vector<int> &deadline, vector<int> &profit) {
        // code here
        int n = profit.size();
        vector<pi> deadprofit(n);
        int maxdeadline = 0;
        for(int i= 0;i<n;i++) {
            deadprofit[i] = {profit[i],deadline[i]};
            maxdeadline = max(maxdeadline,deadline[i]);
        }
        
        int cnt = 0;
        int maxProfit = 0;
        
        sort(deadprofit.begin(),deadprofit.end(),[](pi a, pi b) -> bool {
            if(a.first == b.first)
                return a.second > b.second;
            return a.first > b.first;
        });
        
        
        set<int,greater<int> > mark;
        for(int i = maxdeadline;i>=1;i--) {
            mark.insert(i);
        }
      
        
        for(int i = 0;i<n && !mark.empty();i++) {
            
            int deadline = deadprofit[i].second;
            auto nextBestAvailableSlot = mark.lower_bound(deadline);
            if(nextBestAvailableSlot!=mark.end()) {
                maxProfit+=deadprofit[i].first;
                cnt++;
                mark.erase(nextBestAvailableSlot);
            }
            
            
            
        }
        
        return {cnt,maxProfit};
    }
};

File: Difficulty: Medium/Job Sequencing Problem/job-sequencing-problem.java
class Solution {
    public ArrayList<Integer> jobSequencing(int[] deadline, int[] profit) {
        // code here
        int n = profit.length;
        ArrayList<int[]> arr = new ArrayList<>();
        int maxi = Integer.MIN_VALUE;
        for(int i = 0;i<n;i++) {
            arr.add(new int[] {deadline[i],profit[i]});
            maxi = Math.max(maxi,deadline[i]);
        }
        
        Collections.sort(arr,(a,b)->{
            if(a[1]==b[1]) {
                return Integer.compare(b[0],a[0]);
            }
            return Integer.compare(b[1],a[1]);
        });
        
        // for(int i = 0;i<n;i++) {
        //     System.out.println(arr.get(i)[0]+" "+arr.get(i)[1]);
        // }
        
        ArrayList<Integer> ans = new ArrayList<>();
        
        TreeSet<Integer> st = new TreeSet<>();
        for(int i = 1;i<=maxi;i++) {
            st.add(i);
        }
        int cnt = 0;
        int profits = 0;
        
        for(int i = 0;i<n;i++) {
            int d = arr.get(i)[0];
            int p = arr.get(i)[1];
            
            int index = -1;
            
            try {
                index = st.floor(d);
            } catch (Exception e) {
                index = -1;
            }
            
            
            if(index>0) {
                profits+=p;
                cnt++;
                st.remove(index);
            }
        }
        
        ans.add(cnt);
        ans.add(profits);
        
        return ans;
    }
}